File: content-solutions/queue-deque/deque-window-max.py
from collections import deque


def solution(temps, k):
    window = deque()
    answer = []
    for i, t in enumerate(temps):
        while window and temps[window[-1]] <= t:
            window.pop()
        window.append(i)
        if window[0] <= i - k:
            window.popleft()
        if i >= k - 1:
            answer.append(temps[window[0]])
    return answer

File: content-solutions/queue-deque/deque-window-max.js
function solution(temps, k) {
  const dq = new Array(temps.length);
  let head = 0;
  let tail = 0;
  const answer = [];
  temps.forEach((t, i) => {
    while (tail > head && temps[dq[tail - 1]] <= t) tail -= 1;
    dq[tail++] = i;
    if (dq[head] <= i - k) head += 1;
    if (i >= k - 1) answer.push(temps[dq[head]]);
  });
  return answer;
}
